File: app/src/main/java/com/hana053/micropost/presentation/core/components/micropostlist/PostListAdapter.java
package com.hana053.micropost.presentation.core.components.micropostlist;

import android.support.annotation.Nullable;
import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.ViewGroup;

import com.hana053.micropost.databinding.PostItemBinding;
import com.hana053.micropost.domain.Micropost;
import com.hana053.micropost.presentation.core.components.avatar.AvatarViewListener;
import com.hana053.micropost.presentation.core.components.avatar.AvatarViewModel;

import java.util.ArrayList;
import java.util.Collection;
import java.util.List;

import rx.Observable;

public class PostListAdapter extends RecyclerView.Adapter<PostListAdapter.ViewHolder> {

    private final List<Micropost> posts = new ArrayList<>();

    @Override
    public ViewHolder onCreateViewHolder(ViewGroup parent, int viewType) {
        return ViewHolder.newInstance(parent);
    }

    @Override
    public void onBindViewHolder(ViewHolder holder, int position) {
        final Micropost micropost = posts.get(position);
        final PostItemViewModel viewModel = new PostItemViewModel(micropost);
        holder.binding.setModel(viewModel);
        final AvatarViewModel avatarViewModel = new AvatarViewModel(micropost.getUser());
        holder.binding.setAvatarModel(avatarViewModel);
        holder.binding.setAvatarListener(new AvatarViewListener(avatarViewModel));
    }

    @Override
    public int getItemCount() {
        return posts.size();
    }

    public List<Micropost> getItems() {
        return new ArrayList<>(posts);
    }

    static class ViewHolder extends RecyclerView.ViewHolder {
        private final PostItemBinding binding;

        private ViewHolder(PostItemBinding binding) {
            super(binding.getRoot());
            this.binding = binding;
        }

        private static ViewHolder newInstance(ViewGroup parent) {
            final LayoutInflater inflater = LayoutInflater.from(parent.getContext());
            final PostItemBinding binding = PostItemBinding.inflate(inflater, parent, false);
            return new ViewHolder(binding);
        }
    }

    @Nullable
    public Long getFirstItemId() {
        return Observable.from(posts)
                .take(1)
                .map(Micropost::getId)
                .toBlocking()
                .singleOrDefault(null);
    }

    @Nullable
    public Long getLastItemId() {
        return Observable.from(posts)
                .takeLast(1)
                .map(Micropost::getId)
                .toBlocking()
                .singleOrDefault(null);
    }

    public boolean addAll(int location, Collection<Micropost> posts) {
        if (this.posts.addAll(location, posts)) {
            notifyItemRangeInserted(location, posts.size());
            return true;
        }
        return false;
    }
}

File: app/src/main/kotlin/com/hana053/micropost/shared/posts/PostListAdapter.kt
package com.hana053.micropost.shared.posts

import android.support.v7.widget.RecyclerView
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.ImageView
import android.widget.LinearLayout
import android.widget.TextView
import com.github.curioustechizen.ago.RelativeTimeTextView
import com.hana053.micropost.R
import com.hana053.micropost.domain.Micropost
import com.hana053.micropost.domain.User
import com.hana053.micropost.shared.avatar.AvatarView
import kotlinx.android.synthetic.main.item_posts.view.*
import rx.subjects.PublishSubject


class PostListAdapter(
    private val posts: MutableList<Micropost> = mutableListOf()
) : RecyclerView.Adapter<PostListAdapter.ViewHolder>() {

    val avatarClicksSubject: PublishSubject<User> = PublishSubject.create<User>()

    class ViewHolder(view: View) : RecyclerView.ViewHolder(view) {
        val container: LinearLayout = view.container
        val avatar: ImageView = view.avatar
        val userName: TextView = view.userName
        val createdAt: RelativeTimeTextView = view.createdAt
        val content: TextView = view.content
    }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int) =
        LayoutInflater.from(parent.context)
            .inflate(R.layout.item_posts, parent, false)
            .let(::ViewHolder)

    override fun onBindViewHolder(holder: ViewHolder, position: Int) {
        val item = posts[position]

        holder.apply {
            container.tag = item
            userName.text = item.user.name
            createdAt.setReferenceTime(item.createdAt)
            content.text = item.content
            AvatarView(avatar).render(item.user)
            avatar.setOnClickListener {
                avatarClicksSubject.onNext(item.user)
            }
        }

    }

    override fun getItemCount(): Int = posts.size

    fun getFirstItemId(): Long? = posts.map { it.id }.firstOrNull()

    fun getLastItemId(): Long? = posts.map { it.id }.lastOrNull()

    fun addAll(location: Int, posts: List<Micropost>): Boolean {
        if (this.posts.addAll(location, posts)) {
            notifyItemRangeInserted(location, posts.size)
            return true
        }
        return false
    }


}